packages/medusa/src/services/product.ts

```diff
@@ -0,0 +1,1067 @@
+
+      const { q, query, relations } = this.prepareListQuery_(selector, config)
+
+      const [filteredRelations, shouldLoadStrategyPrices] =
+        this.omitPriceRelationIfExists_(relations)
+
+      if (q) {
+        const [products] = await productRepo.getFreeTextSearchResultsAndCount(
+          q,
+          query,
+          filteredRelations
+        )
+
+        return products
+      }
+
+      const products = await productRepo.findWithRelations(
+        filteredRelations,
+        query
+      )
+
+      return shouldLoadStrategyPrices
+        ? await this.setAdditionalPrices(products, {
+            cart_id: config.cart_id,
+            currency_code: config.currency_code,
+            customer_id: config.customer_id,
+            include_discount_prices: config.include_discount_prices,
+            region_id: config.region_id,
+          })
+        : products
+    })
+  }
+
+  /**
+   * Removes the price relation from a relations array
+   * @param relations relations array from which the prices relation should be removed
+   * @returns a tuple containing the new relations and a boolean indicating whether the price relation was found or not
+   */
+  private omitPriceRelationIfExists_(
+    relations?: string[]
+  ): [string[], boolean] {
+    if (!relations || !relations.length) {
+      return [[], false]
+    }
+    const PRICES_RELATION = "variants.prices"
+
+    return omitRelationIfExists(relations, PRICES_RELATION)
+  }
+
+  /**
+   * Lists products based on the provided parameters and includes the count of
+   * products that match the query.
+   * @param selector - an object that defines rules to filter products
+   *   by
+   * @param config - object that defines the scope for what should be
+   *   returned
+   * @return an array containing the products as
+   *   the first element and the total count of products that matches the query
+   *   as the second element.
+   */
+  async listAndCount(
+    selector: FilterableProductProps | Selector<Product>,
+    config: FindProductConfig = {
+      relations: [],
+      skip: 0,
+      take: 20,
+      include_discount_prices: false,
+    }
+  ): Promise<[Product[], number]> {
+    return await this.atomicPhase_(async (manager) => {
+      const productRepo = manager.getCustomRepository(this.productRepository_)
+
+      const { q, query, relations } = this.prepareListQuery_(selector, config)
+
+      const [filteredRelations, shouldLoadStrategyPrices] =
+        this.omitPriceRelationIfExists_(relations)
+
+      let products
+      let count
+      if (q) {
+        ;[products, count] = await productRepo.getFreeTextSearchResultsAndCount(
+          q,
+          query,
+          filteredRelations
+        )
+      } else {
+        ;[products, count] = await productRepo.findWithRelationsAndCount(
+          filteredRelations,
+          query
+        )
+      }
+
+      if (shouldLoadStrategyPrices) {
```

Comment:
question: Same question as the list method, also, could it be logic to do the same approach in the list method using if instead of the ternary expression? Just for consistency since it is exactly the same approach. also, the else here is unnecessary